### Preprocesado de datos

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("../data_sample/Datos.csv")
df

,provincia,zona,titulo,PrecioActual,PrecioAnterior,metros,habitaciones,ascensor,localizacion,planta,baños,tags,descripcion,Enlace
0,madrid,ciudad-lineal,"Piso en calle de San Marcelo, 22, Ventas, Madrid",355000,0,69,2.0,S,EXTERIOR,5ª,1,"VIVIENDA,LUMINOSO,VISTAS,REFORMADA,PORTERO,EXT...",Particular Vende vivienda totalmente reformada...,https://www.idealista.com/inmueble/106956987/
1,madrid,carabanchel,"Piso en calle Cabo Nicolás Mur, San Isidro, Ma...",149000,159000,91,3.0,N,EXTERIOR,1ª,0,"PISO,EXCLUSIVA,INMOBILIARIA,OPORTUNIDAD",Inmobiliarias Encuentro vende en exclusiva la ...,https://www.idealista.com/inmueble/106906044/
2,madrid,centro,"Piso en calle de Rodas, Lavapiés-Embajadores, ...",195000,0,36,1.0,S,NaN,2ª,0,"EXCLUSIVA,ESTUDIO",ESTUDIO EN PLENO CENTRO DE MADRIDSarago Servic...,https://www.idealista.com/inmueble/107306175/
3,madrid,usera,"Piso en calle de Ferroviarios, Almendrales, Ma...",195000,0,58,1.0,S,INTERIOR,BAJO,0,"VIVIENDA,HOGAR,FUNCIONAL","Esta acogedora vivienda, ubicada en una planta...",https://www.idealista.com/inmueble/106325171/
4,madrid,tetuan,"Dúplex en Bellas Vistas, Madrid",715000,750000,140,3.0,S,EXTERIOR,2ª,0,"TERRAZA,EXCLUSIVA,MODERNO","Maravilloso ATICO de reciente construcción, co...",https://www.idealista.com/inmueble/106627265/
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11821,madrid,centro,"Piso en calle del Amparo, Lavapiés-Embajadores...",189000,0,36,2.0,N,INTERIOR,BAJO,0,"PISO,INMOBILIARIA,HOGAR,NUEVO",¡Descubre tu nuevo hogar en el corazón de Madr...,https://www.idealista.com/inmueble/107292488/
11822,madrid,centro,"Piso en calle Gran Vía, Chueca-Justicia, Madrid",2600000,0,245,2.0,S,EXTERIOR,2ª,0,"PISO,LUJO,EXCLUSIVO,EXTERIOR",Piso totalmente exterior ubicado entre el barr...,https://www.idealista.com/inmueble/103878128/
11823,madrid,tetuan,"Piso en calle del Capitán Blanco Argibay, Vald...",219000,225000,56,1.0,S,EXTERIOR,BAJO,0,"APARTAMENTO,METRO",¡IDEAL INVERSORES! Precioso apartamento en Val...,https://www.idealista.com/inmueble/107025446/
11824,madrid,carabanchel,"Piso en calle de Aceuchal, Vista Alegre, Madrid",165000,0,74,3.0,N,EXTERIOR,2ª,0,"OPORTUNIDAD,METRO","**IDEAL INVERSORES, NO DEJES PASAR ESTA OPORTU...",https://www.idealista.com/inmueble/106946149/


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11826 entries, 0 to 11825
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   provincia       11826 non-null  object 
 1   zona            11826 non-null  object 
 2   titulo          11826 non-null  object 
 3   PrecioActual    11826 non-null  int64  
 4   PrecioAnterior  11826 non-null  int64  
 5   metros          11826 non-null  int64  
 6   habitaciones    11460 non-null  float64
 7   ascensor        11033 non-null  object 
 8   localizacion    10730 non-null  object 
 9   planta          10601 non-null  object 
 10  baños           11826 non-null  int64  
 11  tags            11664 non-null  object 
 12  descripcion     11761 non-null  object 
 13  Enlace          11826 non-null  object 
dtypes: float64(1), int64(4), object(9)
memory usage: 1.3+ MB


In [4]:
df.describe()

,PrecioActual,PrecioAnterior,metros,habitaciones,baños
count,1.182600e+04,1.182600e+04,11826.000000,11460.000000,11826.000000
mean,1.030501e+06,7.359704e+04,153.790039,2.847731,0.394047
std,1.237718e+06,3.639753e+05,766.217750,1.432402,0.882134
min,1.200000e+04,0.000000e+00,11.000000,1.000000,0.000000
25%,2.890000e+05,0.000000e+00,68.000000,2.000000,0.000000
50%,6.200000e+05,0.000000e+00,103.000000,3.000000,0.000000
75%,1.329000e+06,0.000000e+00,160.000000,3.000000,0.000000
max,2.300000e+07,8.450000e+06,75000.000000,20.000000,7.000000


In [5]:
df.isnull().sum().sort_values(ascending=False)

planta            1225
localizacion      1096
ascensor           793
habitaciones       366
tags               162
descripcion         65
metros               0
PrecioAnterior       0
PrecioActual         0
titulo               0
zona                 0
provincia            0
baños                0
Enlace               0
dtype: int64

### Primer análisis

1. COLUMNAS:
- provincia: Provincia del inmueble. Siempre es Madrid (es constante).
- zona: Categórica. Distrito al que pertenece el inmueble.
- titulo: Texto. Título del anuncio.
- PrecioActual: Numérica. Precio de venta actual. En euros. TARGET.
- PrecioAnterior: Numérica. Precio previo en euros.
- metros: Numérica. Superficie en metros cuadrados.
- habitaciones: Numérica. Número de habitaciones.
- ascensor: Categórica binaria. Indica si el inmueble tiene ascensor.
- localizacion: Categórica binaria. Indica la orientación de la vivienda. (exterior o interior)
- planta: Texto (TODO: analizar si puede ser ordinal). Planta del inmueble.
- baños: Numérica. Número de baños.
- tags: Texto en etiquetas separadas por comas. Etiquetas del anuncio.
- descripcion: Texto. Descripción del anuncio.
- Enlace: "Texto". URL del anuncio. Funciona como un ID.

2. LOS DATOS

- Dataset con 11.826 filas y 14 columnas
- Nulos en: planta, localizacion, ascensor, habitaciones, tags, descripcion.
- Los nulos de habitaciones, al ser NaN, pandas lo carga como un float en vez de un int.
- PrecioActual va desde 12.000 a 23.000.000 euros. Media muy desviada (1.030.501) respecto a la mediana (620.000). Asimetría fuerte.
- PrecioAnterior: Media de 73.597, pero los percentiles valen todos 0. Más del 75% de los datos son ceros.
- metros: Media de 154 respecto a la mediana de 103. Desviación estándar de 766. El máximo es 75.000, algo imposible para una vivienda. El mínimo son 11m2, pero ese es, lamentablemente, más razonable.
- habitaciones: de 1 a 20 y mediana 3. El máximo es "plausible".
- baños: mínimo 0, mediana 9 y percentil de 75 en 0 también. Una vivienda no puede tener 0 baños. Hay que investigar.  

Posiblemente los 0s en cualquier columna signifiquen "sin datos". De todas formas, hay que analizarlo y demostrarlo por si no es así.

Procedemos con el estudio de las categóricas. Hay 9 columnas de tipo object donde hay texto o categóricas.

In [6]:
# provincia es una columna constante. El único dato es Madrid. No nos servirá de nada
df.provincia.value_counts()

provincia
madrid    11826
Name: count, dtype: int64

In [ ]:
# Podemos ver 21 zonas distintas. Casi todas las ofertas pertenecen al barrio de salamanca.
# El cual es conocido por ser bastante caro.
# La zona "centro" no parece ser un distrito como los demás.
df.zona.value_counts()

zona
barrio-de-salamanca    2018
centro                 1890
chamberi                796
chamartin               706
tetuan                  596
moncloa                 580
carabanchel             567
ciudad-lineal           537
puente-de-vallecas      526
retiro                  508
hortaleza               444
fuencarral              391
arganzuela              386
san-blas                335
latina                  331
villaverde              314
usera                   302
villa-de-vallecas       212
vicalvaro               193
moratalaz                97
barajas                  97
Name: count, dtype: int64

In [ ]:
# Valores S y N
df.ascensor.value_counts()

ascensor
S    8419
N    2614
Name: count, dtype: int64

In [ ]:
# Llama la atención los valores -1 y -2
df.planta.value_counts()

planta
1ª             2108
2ª             1779
3ª             1570
BAJO           1509
4ª             1266
5ª              816
6ª              574
7ª              289
ENTREPLANTA     163
SÓTANO          128
8ª              124
9ª              106
10ª              47
11ª              35
12ª              22
-1               19
13ª              16
14ª               8
15ª               7
17ª               6
16ª               2
-2                2
20ª               1
27ª               1
22ª               1
21ª               1
18ª               1
Name: count, dtype: int64

In [ ]:
# 57 tags distintos

tags_extraidos = df["tags"].dropna().str.split(",").explode().str.strip() # El explode es para que en el value counts no me busque por listas.
tags_extraidos.value_counts()

tags
PISO                   6115
VIVIENDA               5723
EXTERIOR               2516
METRO                  2237
AMPLIO                 1995
TERRAZA                1922
REFORMADO              1880
OPORTUNIDAD            1832
EXCLUSIVA              1648
LUMINOSO               1436
HOGAR                  1316
ESPECTACULAR           1231
INMOBILIARIA           1205
ÁTICO                  1095
FINCA                  1094
LUJO                    986
VISTAS                  937
NUEVO                   930
EXCLUSIVO               874
REFORMADA               867
EQUIPADA                859
CASA                    851
PARQUE                  800
INTERIOR                792
GARAJE                  792
ESTRENAR                725
PISCINA                 721
FUNCIONAL               663
HALL                    659
MODERNO                 649
APARTAMENTO             638
SUITE                   615
PRESTIGIOSO             587
ELEGANTE                542
IMPRESIONANTE           541
ESTUDIO        

In [ ]:
# Para extraer lo que hay en la columna de titulo hay que ver cómo se estructura y
# casualmente se estructura siempre igual.

# PRIMERO: Tipo inmueble.
# Son 9 tipos distintos
tipo_inmueble = df["titulo"].str.split(" en ").str[0].str.strip()
tipo_inmueble.value_counts()

titulo
Piso                           9551
Ático                           782
Dúplex                          435
Estudio                         339
Casa o chalet independiente     320
Chalet adosado                  213
Chalet pareado                  171
Chalet                           14
Casa rural                        1
Name: count, dtype: int64

In [31]:
# El barrio es un poquito más "tricky", pero sale bien
# Quitamos ", Madrid" del final
sin_madrid = df["titulo"].str.rsplit(",", n=1).str[0]

# El barrio es lo último
barrio = sin_madrid.str.rsplit(",", n=1).str[-1].str.strip()

# Si lleva "X en " (Piso en, Dúplex en... etc), nos quedamos con lo posterior al último " en "
barrio = barrio.str.rsplit(" en ", n=1).str[-1].str.strip()

barrio.value_counts()

titulo
Goya                    612
Recoletos               467
Malasaña-Universidad    451
Castellana              415
Lavapiés-Embajadores    366
                       ... 
El Pardo                  5
Valdecarros               5
Pavones                   5
Aeropuerto                4
Horcajo                   2
Name: count, Length: 139, dtype: int64

### Primeras dos columnas "extra"
Barrio y tipo de inmueble son las dos primeras columnas que irán al dataset.  
Se puede hacer ahora porque es una transformación fila a fila. No hay data leakage, el modelo no entrenaría con esto.

In [32]:
df["tipo_inmueble"] = df["titulo"].str.split(" en ").str[0].str.strip()

df["tipo_inmueble"].value_counts()

tipo_inmueble
Piso                           9551
Ático                           782
Dúplex                          435
Estudio                         339
Casa o chalet independiente     320
Chalet adosado                  213
Chalet pareado                  171
Chalet                           14
Casa rural                        1
Name: count, dtype: int64

In [34]:
# Mismo código que arriba, pero ahora lo metemos en la columna nueva

sin_madrid = df["titulo"].str.rsplit(",", n=1).str[0]
barrio = sin_madrid.str.rsplit(",", n=1).str[-1].str.strip()
df["barrio"] = barrio.str.rsplit(" en ", n=1).str[-1].str.strip()

df["barrio"].value_counts()

barrio
Goya                    612
Recoletos               467
Malasaña-Universidad    451
Castellana              415
Lavapiés-Embajadores    366
                       ... 
El Pardo                  5
Valdecarros               5
Pavones                   5
Aeropuerto                4
Horcajo                   2
Name: count, Length: 139, dtype: int64